### PR TITLE
fix: solana send max amounts

### LINKED ISSUE #8311: Solana send max is wrongly calculating the fee amounts
### Overview

https://explorer.solana.com/tx/66LhMQrzpjudUBV9wKRomC53KAGGmiWUvcLVZSnQw4pDUnTFCgspmSw7dozYVtG14wZnLrhUSqqkizd29p9XzmZF

TX fails upstream when hitting the `Max` button, as we don't have enough compute unit to make it successful

### References and additional details

Spotted in https://github.com/shapeshift/hdwallet/pull/696#pullrequestreview-2488376544

I think we might have a bit more of introspection in the TX instruction, it might use an ATA under the hood or needs more compute units that we actually set (I think peoples usually add a small buffer percent to the compute units so TX are successful most of the time)

### Acceptance Criteria

Can send using the max button

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by NeOMakinG**
## Description

- Solana TXs send was supposing that we had only one instruction, base fees should be multiplied by the number of instructions
- We now take care of the solana minimum rent exemption lamports when calculating the balance in the send modal, so we are never going under this amount, so we never close the account

## Issue (if applicable)

closes #8311

## Risk
High
> High Risk PRs Require 2 approvals

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing
Note to testers, if it's flaky because of compute units, it's mainly because sendMax is false so the extra buffer is not set, you have to spam the sendMax button to have the sendMax as true (https://github.com/shapeshift/web/pull/8328 is supposed to fix it if reviewers agree)

- As it's touching a bit the Solana TX logic, I would suggest to try a few swaps, with non-existing accounts, existing accounts, external accounts with existing ATA or non existing ATA etc
- Try to send the max amount of a solana account
- Try send max for other chains

### Engineering
n/a

### Operations
n/a
- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)
https://jam.dev/c/374e5168-998a-4cbb-8a2d-4cb56e9cd730

### CODE REVIEW

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (NeOMakinG) [inline packages/chain-adapters/src/solana/SolanaChainAdapter.ts:397]: We add a buffer to the compute units, specially for sendMax operations

Other part of the app should be adding the buffer depending on their logic, they might not need it

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (NeOMakinG) [inline packages/chain-adapters/src/solana/SolanaChainAdapter.ts:405]: This is very weird that it worked without this, it's mainly because solana was taking less fees than we thought, but big TXs can revert because of this

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [inline src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:191]: The meaty logic, keep at least the rent exemption amount in the balance so we don't close the account and sendMax is happy

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [inline src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:331]: It was causing an problem, fee estimation was running before this, errored, but the error was kicked because of this effect while we effectively wanted it

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [inline src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:345]: Ok this part I don't like it tbh, I hate using refs for such use cases, it means our pattern is wrong

I used it as a quick win because, when you were using `sendMax` for the first time, the handleInputChange was triggered, and so resetting `sendMax` to false...

Then if you clicked again, as the amount was not changing, it was good 😆

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [inline src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:291]: Don't substract fees if we are not sending native tokens

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/SolanaChainAdapter.ts:405]: Splitted to https://github.com/shapeshift/web/pull/8328

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx]: Splitted to https://github.com/shapeshift/web/pull/8328

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx]: Splitted to https://github.com/shapeshift/web/pull/8328

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (woodenfurniture) [inline src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:285]: [preferably blocking]: Prefer we us the `fromBaseUnit` util here

File: packages/chain-adapters/src/solana/types.ts
Comment (woodenfurniture) [inline packages/chain-adapters/src/solana/types.ts:48]: ❤️

File: packages/chain-adapters/src/solana/constants.ts
Comment (woodenfurniture) [inline packages/chain-adapters/src/solana/constants.ts:6]: [q]: What's the plan for this? Is the intention to leave this in or action it in follow-up?

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (woodenfurniture) [inline packages/chain-adapters/src/solana/SolanaChainAdapter.ts:399]: [nit]: naming

```suggestion
      const instructionCount = instructions?.length || SOLANA_MINIMUM_INSTRUCTIONS_NUMBER
```

File: packages/chain-adapters/src/solana/constants.ts
Comment (woodenfurniture) [inline packages/chain-adapters/src/solana/constants.ts:3]:
```suggestion
export const SOLANA_MINIMUM_INSTRUCTIONS = 3
```

or 

```suggestion
export const SOLANA_MINIMUM_INSTRUCTION_COUNT = 3
```

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (woodenfurniture) [reply packages/chain-adapters/src/solana/SolanaChainAdapter.ts:399]: [preferably blocking]: Ensure the minimum value is never less than `SOLANA_MINIMUM_INSTRUCTIONS_NUMBER`

```suggestion
      const instructionNumber = Math.max(instructions?.length ?? 0, SOLANA_MINIMUM_INSTRUCTIONS_NUMBER)
```

**Review by woodenfurniture — CHANGES_REQUESTED** [review]
First pass, loving the solution here @NeOMakinG . Requesting changes just for the minimum instructions logic (`Math.max`)Will do testing after first batch is actioned 🙏

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (gomesalexandre) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:190]: @NeOMakinG should this live in chain-adapters? We do have `sendMax` option which we leverage in `getFeeData()`, gut feel is this deduction should be part of returned fees?

File: packages/chain-adapters/src/solana/constants.ts
Comment (gomesalexandre) [inline packages/chain-adapters/src/solana/constants.ts:1]: q: Do we have a referenc implementation (e.g backpack) to link as comment doing something similar?

File: packages/chain-adapters/src/solana/types.ts
Comment (gomesalexandre) [reply packages/chain-adapters/src/solana/types.ts:48]: 👋🏽

**Review by gomesalexandre — COMMENTED** [review]
Changeset looking good overall - one q. WRT rent deduction logic not living in chain-adapters, will give it a runtime spin once all comments from big wood are tackled

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:190]: It looks like a good idea but it's actually not, it's not a fee as it's not deducted from the account, so I wouldn't add it to the fees

It might be able to live in `chainSpecific` though, but then it's not so interesting to add it to chain adapter

File: packages/chain-adapters/src/solana/constants.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/constants.ts:6]: > Also, worth commenting where this comes from: https://github.com/coral-xyz/backpack/blob/5a538a41d060d2c48507007f96c766483115aecc/packages/common/src/constants.ts#L433

It's not coming from backpack, but updated the comment to make clear it comes from `getMinimumBalanceForRentExemption`

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (gomesalexandre) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:190]: @NeOMakinG rationale sounds sane to me, indeed it's not *payable* but contributes to the max balance you can send. Only thing is we have to
1. be cognicent that we'll have to add that in many different paths (at the very least, sends and trade execution if e.g sending max to Solana, think that's the only two out the top of my mind for the time being)
2. Comment out what this magic is about

@kaladinlight thoughts? Intuition tells me it feels like an antipattern not to do that in chain-adapters land and feels like overestimating for all other Txs would be a better solution (assuming this is only for max sends), but I think once you go past the intuition solution, the rationale with trading following patterns for correctness actually makes sense

File: packages/chain-adapters/src/solana/constants.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/constants.ts:6]: > [q]: What's the plan for this? Is the intention to leave this in or action it in follow-up?

Honestly I think the effort is a bit big to add another unchained extra call for this, Solana told devs it could change but it never changed for now and everyone is basing their maths on this, I suspect they would at least take some months to update it

Even thought I would feel safer to use the RPC value so we will never have anything to change in the future 

Happy to discuss it in gang gang if needed, we could simply unblock this PR and action a follow up, happy with it too!

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (NeOMakinG) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx:190]: Swapper doesn't have any sendMax notion, and the rent exempt amount is very small, I think users won't hit it, and if they do, adding this to chain adapter if sendMax wouldn't prevent us to make it fail

This is kinda why it shouldn't leave in chain adapter as long as we can't remove this rent exempt from absolutely all txs regardless if it's sendmax, also chain adapter doesn't look like having the notion of the current balance, I feel like it's the responsibility of the consumer to handle this rent logic unfortunately, but happy to see the @kaladinlight input

Let's not block this PR, we can do a follow up anyway! As we would like to release Jupiter and bugfixes!

File: packages/chain-adapters/src/solana/constants.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/constants.ts:4]: Updated in [275f332](https://github.com/shapeshift/web/pull/8327/commits/275f332ecfa01fe8a67b10465016c05190d26415)

File: packages/chain-adapters/src/solana/constants.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/constants.ts]: Updated in [275f332](https://github.com/shapeshift/web/pull/8327/commits/275f332ecfa01fe8a67b10465016c05190d26415)

File: packages/chain-adapters/src/solana/SolanaChainAdapter.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/SolanaChainAdapter.ts]: Ah it's actually a very nice catch, as providing the instruction could be prior the setComputeLimit and setComputeUnitPrice instructions that we can do later in the lifecycle, thanks for this @woodenfurniture 🐐 

Fixed in [275f332](https://github.com/shapeshift/web/pull/8327/commits/275f332ecfa01fe8a67b10465016c05190d26415)

File: packages/chain-adapters/src/solana/constants.ts
Comment (NeOMakinG) [reply packages/chain-adapters/src/solana/constants.ts:11]: > Also, worth commenting where this comes from: https://github.com/coral-xyz/backpack/blob/5a538a41d060d2c48507007f96c766483115aecc/packages/common/src/constants.ts#L433

Updated in [275f332](https://github.com/shapeshift/web/pull/8327/commits/275f332ecfa01fe8a67b10465016c05190d26415)

**Review by gomesalexandre — APPROVED** [review]
Tested locally, confirmed this does what it says on the box:

https://jam.dev/c/237c3c3b-81dc-49a3-b5e6-0697de0c2455

**Review by woodenfurniture — APPROVED** [review]
Can confirm, sol max sends work ✅

Some feedback:
- The gas estimation is cutting REAL fine, so phantom wallet tells me the tx will revert (even though it didnt). Probably a roll of the dice whether it reverts which isnt ideal 🎲
- The gas estimation re-runs on every click of the "max" button, we could cache it probably? See jam for example

https://jam.dev/c/e25ccbd4-9292-4076-bac8-87afd4b11b99

File: src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx
Comment (woodenfurniture) [reply src/components/Modals/Send/hooks/useSendDetails/useSendDetails.tsx]: No that's awesome!

**Pull request merged**
